### ISSUE TITLE
Add Construction of Hadamard matrices up to order 664

### ISSUE THREAD

**Issue by MatteoCati** [sagemath/sage#34807]
<div id="comment:0"></div>

With #34690, all the Hadamard matrices of order <=288 have been added. The next task is to implement matrices of order up to 664 (668 is the first order for which no construction is known). This will require implementing https://core.ac.uk/download/pdf/82189436.pdf, as well as adding some more data for the constructions that are already present.

Depends on #34690

CC:  @dimpase

Component: **combinatorics**

Author: **Matteo Cati**

Branch/Commit: **[`1f624a4`](https://github.com/sagemath/sagetrac-mirror/commit/1f624a44bcd789d9788c693a262699db4313edd8)**

Reviewer: **Dima Pasechnik**

_Issue created by migration from https://trac.sagemath.org/ticket/34807_

**Comment by MatteoCati** [comment]
Commit: **[`7e803fb`](https://github.com/sagemath/sagetrac-mirror/commit/7e803fb32a61a13d289514263c94580814591a5d)**

**Comment by MatteoCati** [comment]
<div id="comment:1"></div>

Last 10 new commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/8eb2285bf4c5d8b11dd3c32b15fd9c52e90112c0"><code>8eb2285</code></a></td><td><code>Add spence construction to general construction function</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/7d9100b1689507f8413b4b9c62f99a605138e6a9"><code>7d9100b</code></a></td><td><code>Add construction for base sequences</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/3ac2cc23b8d8f05b000993c01cc2aaeb47c15383"><code>3ac2cc2</code></a></td><td><code>Generalize construction of T-sequences to work with base sequences</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/733065bf61837714cbfceb4efe9c7a0b764599ed"><code>733065b</code></a></td><td><code>Add Turyn type sequences smallcases</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/369f5131e2146d4f5333524a22a772cf5eb4c282"><code>369f513</code></a></td><td><code>Add construction for small cases of base sequences</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/8ba531eb160e6d42b083ec7a736ae770addc85b9"><code>8ba531e</code></a></td><td><code>Use base sequences in general T-sequences construction</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/129351f629efe82eeaade27dd4f1fe3a01e64c1f"><code>129351f</code></a></td><td><code>Add williamson type matrices of order 7</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/8adf237d807b446dd977d9ecbee1b3e830e52948"><code>8adf237</code></a></td><td><code>Add williamson type matrices of order 9</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/5d329b3ab6e271a96498fdb0f3159fd8110f95c6"><code>5d329b3</code></a></td><td><code>Add construction of skew supplementary difference sets</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/7e803fb32a61a13d289514263c94580814591a5d"><code>7e803fb</code></a></td><td><code>Add construction of skew Hadamard matrices from skew SDS</code></td></tr></table>

**Comment by MatteoCati** [comment]
Branch: **[u/gh-MatteoCati/add_hadamard_matrices_up_to_664](https://github.com/sagemath/sagetrac-mirror/tree/u/gh-MatteoCati/add_hadamard_matrices_up_to_664)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:2"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/cb9b0ca3fed27a9c879034d66cbbe5cb995e183f"><code>cb9b0ca</code></a></td><td><code>fix documentation in difference_family</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`7e803fb`](https://github.com/sagemath/sagetrac-mirror/commit/7e803fb32a61a13d289514263c94580814591a5d)** to **[`cb9b0ca`](https://github.com/sagemath/sagetrac-mirror/commit/cb9b0ca3fed27a9c879034d66cbbe5cb995e183f)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:3"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/b51cc149ad10cc8e0aaee4e8338cc6684e7445c2"><code>b51cc14</code></a></td><td><code>fix seealso directive in T sequences docs</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`cb9b0ca`](https://github.com/sagemath/sagetrac-mirror/commit/cb9b0ca3fed27a9c879034d66cbbe5cb995e183f)** to **[`b51cc14`](https://github.com/sagemath/sagetrac-mirror/commit/b51cc149ad10cc8e0aaee4e8338cc6684e7445c2)**

**Comment by dimpase** [comment]
<div id="comment:4" align="right">comment:4</div>

Wow! That's impressive! Could you say a bit what's in that almost 3000 lines you add here?

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`b51cc14`](https://github.com/sagemath/sagetrac-mirror/commit/b51cc149ad10cc8e0aaee4e8338cc6684e7445c2)** to **[`2cb617c`](https://github.com/sagemath/sagetrac-mirror/commit/2cb617c7e2c9d16ca7f769899dc49b9bfa26b5d3)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:5"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/133e9d16d64fd86219709eb04de66f201e6a62cf"><code>133e9d1</code></a></td><td><code>Fix failing test in base_sequences_smallcases</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/2cb617c7e2c9d16ca7f769899dc49b9bfa26b5d3"><code>2cb617c</code></a></td><td><code>Fix typos in documentations</code></td></tr></table>

**Comment by MatteoCati** [comment]
<div id="comment:6" align="right">comment:6</div>

- I've implemented Spence construction in `hadamard_matrix.py`. This uses the supplementary difference sets created by `supplementary_difference_set` in `difference_faimilies.py`. In the same file, the relative difference sets used by Spence to construct supplementary difference sets are created by the function `relative_difference_set`.
- In `t_sequences.py`, I've added some new T-sequences. These are constructed from base sequences using the function `T_sequences_construction_from_base_sequences` (this function was already present, but was being used only for the case when p=1). Also, base sequences are constructed from Turyn type sequences using the (new) function `base_sequences_construction` (I've also added the function `is_base_sequences_tuple`, to check that the construction is correct). Turyn type sequences are stored directly in `turyn_type_sequences_smallcases`.
- Lastly, I've added in `difference_family.py` the function `skew_supplementary_difference_set` which creates skew SDS (as described by  Djokovic in https://ajc.maths.uq.edu.au/pdf/10/ocr-ajc-v10-p259.pdf), which are used by `GS_skew_hadamard_smallcases` in `hadamard_matrix.py` to  create new skew Hadamard matrices.

Also, note that the older commits (up until bd025a2e87389b9b7f5573049ad62cdd087e2bf0) are coming from ticket #34690.

**Comment by dimpase** [comment]
<div id="comment:7" align="right">comment:7</div>

Does this mean that we also have new (for Sage) orders for skew Hadamard matrices?

**Comment by MatteoCati** [comment]
<div id="comment:8" align="right">comment:8</div>

Yes, I've added skew Hadamard matrices of order 268, 412, 452, 508, 604, 628, 652, 724, 964.

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:9"></div>

Branch pushed to git repo; I updated commit sha1. This was a forced push. Last 10 new commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/e231e5e03e5ef43b2d7d922d0b1cb192a9a8befb"><code>e231e5e</code></a></td><td><code>Add construction for small cases of base sequences</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/fcdcbb7a2d0b2dfea86fda40309b88507f1aef4f"><code>fcdcbb7</code></a></td><td><code>Use base sequences in general T-sequences construction</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/df544a0ad8fb6a4d7556bec73cef35fc72f11f68"><code>df544a0</code></a></td><td><code>Add williamson type matrices of order 7</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/35fffdabe13661c6629e866172cf8d96e426a500"><code>35fffda</code></a></td><td><code>Add williamson type matrices of order 9</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/809783006e148c1cd13e72527ff369b94fbcf37f"><code>8097830</code></a></td><td><code>Add construction of skew supplementary difference sets</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/fbc27fc5633ea6084038177ff34c595829abf38a"><code>fbc27fc</code></a></td><td><code>Add construction of skew Hadamard matrices from skew SDS</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/7a9e0025ee3dbf89a8af0b90a2b32f061f39e3de"><code>7a9e002</code></a></td><td><code>fix documentation in difference_family</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/eb3474de2516c56dd4eac90f18fde3c9d05fde96"><code>eb3474d</code></a></td><td><code>fix seealso directive in T sequences docs</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/3d6f3fb5f0f77a7e0de494f6f30dc49896c4063d"><code>3d6f3fb</code></a></td><td><code>Fix failing test in base_sequences_smallcases</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/1f624a44bcd789d9788c693a262699db4313edd8"><code>1f624a4</code></a></td><td><code>Fix typos in documentations</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`2cb617c`](https://github.com/sagemath/sagetrac-mirror/commit/2cb617c7e2c9d16ca7f769899dc49b9bfa26b5d3)** to **[`1f624a4`](https://github.com/sagemath/sagetrac-mirror/commit/1f624a44bcd789d9788c693a262699db4313edd8)**

**Comment by MatteoCati** [comment]
<div id="comment:10" align="right">comment:10</div>

Rebased over latest beta.

**Comment by dimpase** [comment]
<div id="comment:11" align="right">comment:11</div>

Replying to [Matteo Cati](#comment%3A8):
> Yes, I've added skew Hadamard matrices of order 268, 412, 452, 508, 604, 628, 652, 724, 964.

nice. 

How about bringing Sage up to date w.r.t. skew Hadamard matrices too?
I actually don't know what the 1st unknown order is in this case (something that's probably mentioned in the literature, I didn't try too hard to find this),
but I noticed that Sage does not know about orders 76, 100, and 116, whereas examples are known, see e.g.

https://mathscinet.ams.org/mathscinet-getitem?mr=2416980

**Comment by dimpase** [comment]
Reviewer: **Dima Pasechnik**

**Comment by MatteoCati** [comment]
<div id="comment:14" align="right">comment:14</div>

Replying to [Dima Pasechnik](#comment%3A11):
> How about bringing Sage up to date w.r.t. skew Hadamard matrices too?

Okay, I will open a new ticket (#34848).

**Comment by vbraun** [comment]
Changed branch from **[u/gh-MatteoCati/add_hadamard_matrices_up_to_664](https://github.com/sagemath/sagetrac-mirror/tree/u/gh-MatteoCati/add_hadamard_matrices_up_to_664)** to **[`1f624a4`](https://github.com/sagemath/sagetrac-mirror/commit/1f624a44bcd789d9788c693a262699db4313edd8)**
